File: Problem2-Diagonal_order.py
class Solution(object):
    def findDiagonalOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        #base condition
        if not matrix:
            return []
        #direction variable (+1=up, -1=down)
        dir=1
        #boundaries
        i=0
        j=0
        r=len(matrix)
        c=len(matrix[0])
        res=[]
        #loop end condition>> until you get all elements in result array
        while (len(res)!=(r)*(c)):
            #append value to result>>starter res=[matrix[0][0]]
            res.append(matrix[i][j])
            #check conditions and move on
            if dir==1:
                #order of if and elif very important
                #consider element 3, if we increment j it will go out of bounds 
                #(when we consider (i==0) condition as first if block)
                if (j==c-1):
                    dir=-1
                    i+=1
                elif(i==0):
                    dir=-1
                    j+=1
                else:
                    i-=1
                    j+=1
                    
            elif dir==-1:
                if(i==r-1):
                    dir=1
                    j+=1
                elif(j==0):
                    dir=1
                    i+=1
                else:
                    i+=1
                    j-=1
        return res
                


#O(m*n)
#O(1) as result array was supposed to be returned so no auxiliary array
        
        

File: Problem2_DiagonalOrder.cs
// Time Complexity : O(m*n)
// Space Complexity : O(m*n)
// Did this code successfully run on Leetcode : Yes
// Starting from the first element of the matrix mat, we traverse through elements in upward direction (direction =true, decrement row and increment column)
// until reaching the last column or first row, then change the direction to downwards (direction = false). Similarly traverse through the elements in downward 
// direction (direction = false, i++, j--) until reaching the last row or the first column, then change the direction to upwards.
//S tore the traversed element in the result array until all the elements are traversed.
public class Solution {
    public int[] FindDiagonalOrder(int[][] mat) {
        int m = mat.Length;
        int n = mat[0].Length;
        int[] result = new int[m*n];
        bool direction = true;
        int i = 0;
        int j = 0;
        result[0] = mat[i][j];
        for(int k = 1; k < (m*n); k++) {
            if(direction) {
                if(j == n-1) { // On reaching the last column of top row, change the direction to downwards and move to next row
                    i++;
                    direction = false;
                }
                else if(i == 0) { //On reaching top row, change the direction to downwards and move to next column
                    j++;
                    direction = false;
                }
                else { //Until reaching the top row or last column, keep moving upwards
                    i--; j++;
                }
            }
            else {
                if(i == m-1) {//On reaching last row, change the direction to upwards and move to next column
                    j++; direction = true;
                }
                
                else if(j == 0) {//On reaching first column, change the direction to upwards and move to next row
                    i++; direction = true;
                }
                else {//Until reaching the first column or last row, keep moving downwards
                    i++; j--;
                }
            }
            result[k] = mat[i][j];
        }
        return result;
    }
}
